src/rate.rs

```diff
@@ -0,0 +1,218 @@
+use anyhow::{bail, Context};
+use num::{BigUint, Integer, ToPrimitive};
+use std::convert::TryFrom;
+use std::iter::FromIterator;
+use std::str::FromStr;
+
+/// Represent a rate. Note this is designed to support Bitcoin/Dai buy and sell rates (Bitcoin being in the range of 10k-100kDai)
+/// A rate has a maximum precision of 9 digits after the decimal
+// rate = self.0 * 10e-9
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct Rate(u64);
+
+impl Rate {
+    pub const PRECISION: u16 = 9;
+    const MAX_RATE: u64 = (u64::MAX / 10) ^ 9;
+
+    /// integer = rate * 10ePRECISION
+    pub fn new(integer: u64) -> anyhow::Result<Self> {
+        if integer > Self::MAX_RATE {
+            bail!("Rate is larger than supported");
+        }
```

Comment:
Could a smaller input data type fulfill the same purpose? Like a u32, or a u16?

Reply:
I think I am over-engineering this, will remove the check.